Add testing guide - introduction / overview

vonovak opened the pull request:
I'm putting up first version of testing guide for RN apps. Since it was written by just me (and @rachelnabors provided review and did some improvements and corrections) we'll need more people looking at this and providing feedback before it can be published.

This part of the testing doc is an introduction, more theoretical than practical and can be thought of as the RN counterpart of [react testing overview](https://reactjs.org/docs/testing.html). The plan is to add a more example-driven doc later, kinda like [react testing recipes](https://reactjs.org/docs/testing-recipes.html).

[preview here](https://deploy-preview-1469--react-native.netlify.com/docs/next/testing)

Thanks for your comments!

TheSavior commented on docs/testing.md:
```diff
@@ -0,0 +1,179 @@
+
+Similarly, with software, writing your code in many small functions that you can test (like adding sensors to the different components of an engine) is more effective than writing your entire program in one file with many lines of code (where you can only test if the engine starts or not). In this way, writing testable code is intertwined with writing clean, modular code. We encourage you to expl ...
+
+To help make your app more testable, it’s a good idea to separate the view part of your app, React components, from business logic and app state (regardless of whether you use Redux, MobX or other solutions). This way, you can separate tests of your business logic — which shouldn’t rely on your React components — from the components themselves, whose job is primarily rendering your app’s UI!
```
Is there another post online somewhere that we like that we can link to for how to write testable React / JS code?

vonovak replied:
I did some googling, but didn't find too much content. Would you like to add some of these?

https://www.toptal.com/javascript/writing-testable-code-in-javascript  
https://testing.googleblog.com/2008/08/by-miko-hevery-so-you-decided-to.html?m=1 
https://www.toptal.com/qa/how-to-write-testable-code-and-why-it-matters 
https://dzone.com/articles/write-testable-code

rachelnabors replied:
I endorse you writing your own post... with us :3

TheSavior commented on docs/testing.md:
````diff
@@ -0,0 +1,179 @@
+});
+```
+
+The test is described by the string passed to the [`it`](https://jestjs.io/docs/en/api#testname-fn-timeout) function - take a good care writing the description so that it’s clear what is being tested. Do your best to cover the following:
````
s/take a good care/take good care/

vonovak replied:
fixed

TheSavior commented on docs/testing.md:
```diff
@@ -0,0 +1,179 @@
+
+### Structuring Tests
+
+Your tests should be short and ideally test only one thing. Let's start with an example unit test:
```
Should this say something like "start with an example unit test written with Jest"?

vonovak replied:
did that

TheSavior commented on docs/testing.md:
```diff
@@ -0,0 +1,179 @@
+
+## Unit tests
+
+The lowest level at which you can test is unit testing. Unit tests cover the smallest parts of code, like individual functions or classes.
```
Is "The lowest level at which you can test is unit testing" a phrase that will make sense? Is this sentence necessary or is the following sentence enough?

vonovak replied:
I think just the second sentence is enough, I removed the first.

TheSavior commented on docs/testing.md:
```diff
@@ -0,0 +1,179 @@
+>
+> Then it falls into the “integration test” category.
+
+## Component Tests
```
Is this supposed to be a sub point of Integration Tests? It is at the same heading level

vonovak replied:
I'm myself not sure. I think component tests can fall below unit tests as well as integration tests, so I covered it separately, also because it's a little different (requires extra deps) from testing vanilla js.

vonovak replied:
I addressed this by saying "Component tests could fall into both unit and integration testing, and since they are such a core part of React Native, we'll cover them in separate paragraphs."

TheSavior commented on docs/testing.md:
````diff
@@ -0,0 +1,179 @@
+}
+```
+
+When testing user interactions, test the component from the user perspective: you could write your tests to call `addNewItemToShoppingList` directly, and while such test works, it’s very close to the implementation details of your component.
````
s/it’s very close to the implementation details of your component/it would break if you refactor the implementation details of your component/

vonovak replied:
I reworded that to 

> and while such test works, it's very close to the implementation details of your component and would break if you refactor the component.

hope that works

TheSavior commented on docs/testing.md:
```diff
@@ -0,0 +1,179 @@
+
+[Snapshot testing](https://jestjs.io/docs/en/snapshot-testing) lets you test how—and if—a component is rendering. A snapshot is a textual representation of your component’s output that gets committed in your repo. When using snapshot testing, you first implement your component and then run the snapshot test, which saves the snapshot to a file. Any future changes to the component will result in a  ...
+
+In theory, you can use snapshots to test anything that is serializable, but do not overuse them! Snapshot is created _after_ you’re done working on the component, and at that point the snapshot is considered to be correct (even in the case when the rendered output is actually wrong). You need to make sure that the snapshot contains what you expect (which may not be trivial because snapshots can b ...
```
Typo at "Snapshot is created" Does it need a "The"?

vonovak replied:
Sorry, I'm not sure what "The" you're referring to, but I have checked " Allow edits from maintainers." so maybe you can edit it right here in the PR.

vonovak replied:
note that I expanded the snapshot testing paragraph in https://github.com/facebook/react-native-website/pull/1469/commits/127218a09cec6f58bcd4039825e567982eae64f0

rachelnabors replied:
@vonovak We are using hooks and function components in the docs exclusively going forward. If you can do this, I can merge!

vonovak replied:
snapshot part was rewritten with help of thymikee and class component is gone (example uses hooks) so resolving.

rachelnabors commented on docs/testing.md:
```diff
@@ -0,0 +1,179 @@
+
+## Writing Testable Code
+
+To start with tests, you first need to write code that is testable. Consider a car’s engine. For some reason it won’t start. In older cars, figuring out what is wrong with the engine requires the mechanic to take the engine apart, piece by piece, until they can figure out the issue. In modern cars, engines are outfitted with sensors that keep an eye on parts of the system, determining the exact l ...
```
I wrote this bit ;p Often times looping in a technology folks already know and are familiar with to paint a picture makes the information more sticky.

rickhanlonii replied:
I like that we're using a metaphor to explain tests!

I think there's a difference between the way sensors in cars work and the way code tests work because sensors do not run parts of the engine like tests do. Sensors are more like static analysis in the above section. Is there a better metaphor for this section we could use?

vonovak replied:
How would this sound?

To start with tests, you first need to write code that is testable. Consider an aircraft manufacturing process - before a model first takes off, individual parts are tested to guarantee they are safe and function correctly. For example, the durability of wings is tested by bending them under extreme load, engine parts are tested for their durability, windshield is tested against simulated bird impact, and much more.

rachelnabors replied:
I endorse this new metaphor! Airplanes are pretty cool, too!

vonovak replied:
ok, I'm considering this resolved then

vonovak commented:
@TheSavior @rickhanlonii @rachelnabors I updated the docs, hopefully resolving some of the comments. I also changed the car engine example to something else, but not sure if it's better 😄 Please let me know :)

thymikee commented on docs/testing.md:
```diff
@@ -0,0 +1,204 @@
+
+After writing testable code, it’s time to write some actual tests! (Or if you do test-driven development, you actually write tests first!). We encourage you to explore the topic of testability in other learning resources.
+
+React Native apps come with [Jest](https://jestjs.io) out of the box, ready to use. Jest is a popular JavaScript Testing Framework also started at Facebook. You can use it to write all types of tests featured in this guide. Let’s get started!
```
```suggestion
The default template of React Native ships with [Jest](https://jestjs.io) testing framework. It includes a preset that's tailored to this environment so you can get productive without tweaking the configuration and mocks straight away. You can use Jest to write all types of tests featured in this guide. Let’s get started!
```

vonovak commented on docs/testing.md:
```diff
@@ -0,0 +1,204 @@
+
+## End to End Tests
+
+In End-to-end (E2E) tests, you verify your app is working on a device from the user perspective.
```
@thymikee I find the word "environment" to be too ambiguous, I'll write real device or simulator if you don't mind.

thymikee replied:
Sure, appreciated!

vonovak commented on docs/testing.md:
```diff
@@ -0,0 +1,204 @@
+
+Instead, E2E testing libraries allow you to find and control elements in the screen of your app: for example, you can tap buttons or insert text into textinputs the same way a real user would. Then you can make assertions about whether or not certain element exists in the app’s screen, whether or not it’s visible, what text it contains and so on.
+
+E2E tests give you the highest possible confidence that part of your app is working but the tradeoff is that writing and running them is _much_ more time consuming than with the previously mentioned types of tests. E2E tests also are more demanding when it comes to their maintenance.
```
thanks, I'll rephrase slightly because I want to state that writing them is also harder compared to the other test types

vonovak commented on docs/testing.md:
````diff
@@ -0,0 +1,204 @@
+}
+```
+
+When testing user interactions, test the component from the user perspective: you could write your tests to call `addNewItemToShoppingList` directly, and while such test works, it's very close to the implementation details of your component and would break if you refactor the component.
````
I actually think that the original has a slightly better "flow" but I'm not a native speaker. @rachelnabors what do you think? Thank you.

thymikee replied:
I'm only following learnings from http://www.hemingwayapp.com/ which suggests to increase readability by constructing less complex sentences

rachelnabors replied:
Ha, and interestingly enough, complex sentences are harder for translation services and non-native English speakers to muddle through. I think it's better as @vonovak originally wrote it!

vonovak commented on docs/testing.md:
````diff
@@ -0,0 +1,204 @@
+Aside from rendering some UI, your components handle events like `onChangeText` for `TextInput` or `onPress` for `Button`. They may also contain other functions and event callbacks. Consider the following example:
+
+```js
+class GroceryShoppingListCreator extends React.Component {
````
I'll see what I can do, I want this

> you could write your tests to call `addNewItemToShoppingList` directly, and while such test works, it's very close to the implementation details of your component and would break if you refactor the component.

to make sense and if I add hooks, things will become a little less straightforward imho

rachelnabors replied:
@vonovak We are using hooks and function components in the docs exclusively going forward. If you can do this, I can merge!

rachelnabors replied:
I know it will be a little less obvious, but it is/will be the more widespread usecase, meaning investing in conveying this right with hooks will pay off more than sticking with the easier explanation in classes.

vonovak replied:
@rachelnabors I rewrote the example with hooks, and I think it's for the better.

vonovak commented:
@thymikee firstly, thanks for your comments! I think I resolved most of them and will take another look at the rest later this week.

> I'm especially concerned with "snapshot testing" because it's vastly misunderstood and misused feature in React projects. I'd even consider not mentioning it at all, and later amend this article with something suitable 

> ... my experience is that people will overuse it, because Jest exposes too low-level API (the toMatchSnapshot).

> What we should focus on is to how to write a good snapshot test – which is, using libs like snapshot-diff and ESLint rules like no-large-snapshots.

I think snapshot testing should be mentioned here in _some_ form because people will expect it to be covered here. There's a full article on snapshot testing in the jest docs https://jestjs.io/docs/en/snapshot-testing that I guess you're aware of. Also there is this page: https://jestjs.io/docs/en/tutorial-react-native that talks a lot about snapshots (perhaps that page should link here.) So question is how to handle this right. Perhaps we could 

- keep a short paragraph here about snapshot testing and the advice "When in doubt, prefer explicit expectations as described in the previous paragraph."

- update the snapshot testing article in jest docs and we could link that article from here?

thymikee commented:
Yea, I hate these 2 pages mentioning snapshots this way :P. I think it would be nice to talk _just a bit_ about snapshot, and show an example with `snapshot-diff` only, without using `toMatchSnapshot`. That would address my initial concerns :)

vonovak commented:
@thymikee 

> I think it would be nice to talk just a bit about snapshot.

I agree, there are entire pages written just about snapshots, as linked above. So I'll try to keep it short.

> show an example with `snapshot-diff` only, without using `toMatchSnapshot`

Personally, I think it's best to be honest - if something is easy to be misused / overused, let's state that explicitly, not talking about it does not help, especially given that the internet is full of articles on snapshot testing.

thymikee commented:
Let's do it then! Really appreciate the work you put into this guide, it's so important to have it.

vonovak commented:
@thymikee I updated the paragraph, please let me know if it's okay with you. It's actually a little longer than before 😅 but I feel like it covers what needs to be said.

>  show an example with snapshot-diff only, without using toMatchSnapshot

I did not include any example to keep this short, but my plan is to create another, more practical guide, see my first https://github.com/facebook/react-native-website/pull/1469#issue-341740368;

thymikee commented on docs/testing.md:
````diff
@@ -0,0 +1,221 @@
+</Text>
+```
+
+Snapshots are _generated_ by a component testing testing library, unlike the other types of tests which are written manually (snapshots are too complex to be created by hand).
````
Let's remove this paragraph

thymikee replied:
If you don't want to remove it, please at least remove the `testing testing` duplication, I forgot to mention earlier 😅

vonovak replied:
@thymikee I did remove it in https://github.com/facebook/react-native-website/pull/1469/commits/4a04c6877333907f9e699af53915a68c0c6bc105

thymikee commented on docs/testing.md:
```diff
@@ -0,0 +1,221 @@
+
+### Testing Rendered Output
+
+[Snapshot testing](https://jestjs.io/docs/en/snapshot-testing) lets you verify how component is rendering. A snapshot is a textual representation of your component’s render output, and may look like this:
```
What do you think about this phrasing?
```suggestion
[Snapshot testing](https://jestjs.io/docs/en/snapshot-testing) is a special kind of testing enabled by Jest. It is a very powerful and low-level tool, so special attention is advised when using it. 

> In this guide we will speak about "component snapshot", which is a snapshot specialized to render React component tree objects using a JSX-like string, powered by a [custom React serializer](https://jestjs.io/docs/en/configuration#snapshotserializers-arraystring) that lets Jest translate JS objects to string that's human-readable. It's built into Jest and used by default so you don't need to worry about it.

A component snapshot is a textual representation of your component’s render output _generated_ during a test run, and may look like this:
```

vonovak replied:
@thymikee I took the suggestion but not sure about it upon re-reading it: If I simplify what's written there, it sounds like "component snapshot is a snapshot specialized to render React components using a string",  which sounds a little odd.

here's the text for reference:

> In this guide we will speak about "component snapshot", which is a snapshot specialized to render React component tree objects using a JSX-like string, powered by a custom React serializer that lets Jest translate JS objects to string that's human-readable. 

How about this?

In this guide we will speak about "component snapshot", which is a JSX-like string, created by a custom React serializer that lets Jest translate React component trees to string that's human-readable.

thymikee replied:
Sounds good 👍

thymikee commented on docs/testing.md:
```diff
@@ -0,0 +1,221 @@
+
+Snapshots are _generated_ by a component testing testing library, unlike the other types of tests which are written manually (snapshots are too complex to be created by hand).
+
+With snapshot testing, you typically first implement your component and then run the snapshot test, which creates a snapshot and saves it to a file in your repo as a reference snapshot. The file is then committed and checked during code review. Changes to the component render output will result in change of its snapshot, which fails the test. You then need to update the stored reference snapshot  ...
```
```suggestion
With snapshot testing, you typically first implement your component and then run the snapshot test, which creates a snapshot and saves it to a file in your repo as a reference snapshot. **The file is then committed and checked during code review**. Changes to the component render output will result in change of its snapshot, which fails the test. You then need to update the stored reference snapshot for the test to pass. That change again needs to be committed and reviewed.
```

vonovak commented on docs/testing.md:
```diff
@@ -0,0 +1,204 @@
+
+There is more value in testing, that perhaps isn't immediately visible: when there is a bug in your code, often the easiest way to fix it is to write a failing test that exposes the bug - then when you fix the bug in your code and re-run the test, it'll pass, meaning the bug is fixed and is never reintroduced into the code base.
+
+Tests can also serve as documentation for new people joining your team - when they have to use code they have never seen before, reading tests can greatly help them to understand how the existing code is intended to be used.
```
I moved this into separate "motivation paragraph"

vonovak commented on docs/testing.md:
```diff
@@ -0,0 +1,204 @@
+
+## Summary
+
+Testing is important because it proves that your code is working, and perhaps even more importantly, ensures that your code continues to work in the future as you add new features, refactor the existing ones, or after you upgrade major dependencies of your project.
```
used the suggestion so resolving

vonovak commented on docs/testing.md:
```diff
@@ -0,0 +1,204 @@
+      <>
+        <TextInput
+          placeholder="Enter grocery item"
+          onChangeText={this.setGroceryItem}
```
example is now valid code

vonovak commented:
@rachelnabors at this point, a couple of people already provided some review, and I feel like most of the findings were addressed, so this should be good for merge. If you want someone to do another round of review, I'll be ready :)

Also tagging @TheSavior as he provided some comments - please take a look if they are ok to be resolved, thanks!

rachelnabors commented:
Re: [Jest](https://github.com/facebook/react-native-website/pull/1469#issuecomment-578939331), if someone submits a PR, I can get it merged! 

@vonovak Are you still waiting on feedback for the remaining unresolveds?

vonovak commented:
@rachelnabors I rewrote the example so it uses hooks. It's better I think 🙂 It'd be nice if someone could take a look at it as my hooks-fu is still not great.
I left 1 comment unresolved. I don't think it's a blocker though.
